csrc/multidevice/communication.cpp

```diff
@@ -111,209 +98,225 @@ Broadcast::Broadcast(CommParams params) : Communication(params, "broadcast") {}
     Communicator& comm,
+    at::Tensor input_tensor,
+    at::Tensor output_tensor,
     std::optional<CommunicatorBackend> backend) {
   post_common(*this, comm);
-  // This is used to change the representation of the buffers to match c10d
-  // ProcessGroup API
-  std::vector<std::vector<at::Tensor>> buf_list;
-  if (comm.deviceId() == params_.root) {
-    assertBufferCount(params_.src_bufs, params_.team.size());
-    buf_list = {std::move(params_.src_bufs)};
-  } else {
-    assertBufferCount(params_.src_bufs, 0);
+
+  if (comm.deviceId() == params_.root && !params_.is_root_in_mesh) {
+    output_tensor = at::empty_like(input_tensor.slice(0, 0, 1));
   }
-  auto work =
-      comm.getBackendForTeam(params_.team, backend)
-          ->scatter(
-              params_.dst_bufs, buf_list, {.rootRank = root_relative_index_});
+  std::vector<at::Tensor> output_tensors({output_tensor});
+
+  std::vector<std::vector<at::Tensor>> input_tensors;
   if (comm.deviceId() == params_.root) {
-    params_.src_bufs = std::move(buf_list.back());
+    input_tensors.resize(1);
+    int64_t j = 0;
+    for (auto i : c10::irange(params_.team.size())) {
+      if (root_relative_index_ == static_cast<DeviceIdxType>(i) &&
+          !params_.is_root_in_mesh) {
+        input_tensors.front().push_back(output_tensor);
+        continue;
+      }
+      input_tensors.front().push_back(input_tensor.slice(0, j, j + 1));
+      j++;
+    }
   }
-  return work;
-}
 
-Reduce::Reduce(CommParams params) : Communication(params, "reduce") {
-  assertBuffersHaveSameSize(params_.src_bufs, params_.dst_bufs);
-  assertBufferCount(params_.src_bufs, 1);
+  return comm.getBackendForTeam(params_.team, backend)
+      ->scatter(
+          output_tensors, input_tensors, {.rootRank = root_relative_index_});
 }
 
+Reduce::Reduce(CommParams params) : Communication(params, "reduce") {}
+
 c10::intrusive_ptr<c10d::Work> Reduce::post(
     Communicator& comm,
+    at::Tensor input_tensor,
+    at::Tensor output_tensor,
     std::optional<CommunicatorBackend> backend) {
+  post_common(*this, comm);
+
+  at::Tensor tensor;
   if (comm.deviceId() == params_.root) {
-    assertBufferCount(params_.dst_bufs, 1);
+    if (params_.is_root_in_mesh) {
+      doLocalCopy(output_tensor, input_tensor);
+      tensor = output_tensor;
+    } else {
+      NVF_ERROR(
+          output_tensor.scalar_type() == at::kFloat,
+          "only float tensors are supported");
+      output_tensor.fill_(getInitialValue<float>(params_.redOp));
+      tensor = output_tensor;
+    }
   } else {
-    assertBufferCount(params_.dst_bufs, 0);
+    tensor = input_tensor;
   }
-  post_common(*this, comm);
-  auto& buf =
-      (comm.deviceId() == params_.root) ? params_.dst_bufs : params_.src_bufs;
+  std::vector<at::Tensor> tensors({tensor});
+
   c10d::ReduceOptions options = {
       .reduceOp = params_.redOp, .rootRank = root_relative_index_};
-  auto team_backend = comm.getBackendForTeam(params_.team, backend);
```

Comment:
why removing this?

Reply:
I'll add it back in a follow-up PR. I found that removing reduce_oop helped me a lot understand the code, so my first version was made that way. However, I do understand the speed benefit of using OOP reduction when it's available.